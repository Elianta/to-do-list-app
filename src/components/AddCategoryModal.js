import React from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import Modal from 'react-modal';
import {closeModals} from '../actions/modals';
import {addNestedCategory} from '../actions/categories';

export class AddCategoryModal extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            category: ''
        };
        this.handleSaveNewCategory = this.handleSaveNewCategory.bind(this);
    }

    handleSaveNewCategory(e) {
        e.preventDefault();
        const categoryName = this.state.category;
        this.props.addNestedCategory({name: categoryName, parentID: this.props.modals.editableCategory.id});
        this.props.closeModals();
    }

    render() {
        return (
            <Modal
                isOpen={this.props.modals.isAddCategoryModalOpen}
                onRequestClose={this.props.handleModalClose}
                contentLabel="Add Category"
                closeTimeoutMS={200}
                className="modal"
            >
                <h3 className="modal__title">{`Add nested category into '${this.props.modals.editableCategory.name}'`}</h3>
                <form data-testid="form" className="modal__form" onSubmit={this.handleSaveNewCategory}>
                    <input
                        type="text"
                        name="category"
                        autoFocus={true}
                        className="modal__input"
                        data-testid="name"
                        onChange={(e) => {
                            this.setState({category: e.currentTarget.value.trim()});
                        }}
                    />
                    <button className="button  modal__btn">Save</button>
                </form>
            </Modal>
        );
    }
}

const mapStateToProps = (state) => ({
    modals: state.modals
});

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
        addNestedCategory,
        closeModals
    }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(AddCategoryModal);
