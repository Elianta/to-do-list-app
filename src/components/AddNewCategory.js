import React from 'react';
import {connect} from 'react-redux';
import {addCategory} from '../actions/categories';

class AddNewCategory extends React.Component {
    constructor(props) {
        super(props);
        this.handleAddNewCategory = this.handleAddNewCategory.bind(this);
        this.state = {
            error: undefined
        };
    }

    handleAddNewCategory(e) {
        e.preventDefault();
        const categoryName = e.target.elements.category.value.trim();
        if (!categoryName) {
            this.setState(() => ({error: 'Enter valid value to add item'}));
        } else {
            this.setState(() => ({
                error: ''
            }));
            this.props.dispatch(addCategory(categoryName));
            e.target.elements.category.value = '';
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form className="add-item-form" onSubmit={this.handleAddNewCategory}>
                    <input
                        className="add-item-form__input"
                        type="text"
                        name="category"
                        placeholder="Enter category title..."
                    />
                    <button className="button">Add</button>
                </form>
            </div>
        );
    }
}

export default connect()(AddNewCategory);
