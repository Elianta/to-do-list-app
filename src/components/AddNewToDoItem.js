import React from "react";

export default class AddNewToDoItem extends React.Component {
  constructor(props) {
    super(props);
    this.handleAddNewToDoItem = this.handleAddNewToDoItem.bind(this);

    this.state = {
      error: undefined
    };
  }

  handleAddNewToDoItem(e) {
    e.preventDefault();
    const toDoItem = e.target.elements.toDoItem.value.trim();
    const error = this.props.handleAddNewToDoItem(this.props.categoryID, toDoItem);

    this.setState(() => ({ error }));

    if (!error) {
      e.target.elements.toDoItem.value = '';
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form className="add-item-form" onSubmit={this.handleAddNewToDoItem}>
          <input className="add-item-form__input" type="text" name="toDoItem" placeholder="Enter item title"/>
          <button className="button">Add</button>
        </form>
      </div>
    )
  }
}
