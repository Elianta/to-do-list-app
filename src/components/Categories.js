import React from 'react';
import Category from './Category.js';

class Categories extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            categoryToMoveTaskInId: null
        };
        this.handleSpecifyCategoryToMove = this.handleSpecifyCategoryToMove.bind(this);
    }

    handleSpecifyCategoryToMove(categoryID) {
        this.setState(() => ({
            categoryToMoveTaskInId: categoryID
        }));
        this.props.handleSpecifyCategoryToMove(categoryID);
    }

    render() {
        let categories = this.props.categories.map((category) => {
            return (
                <Category
                    node={category}
                    children={category.children}
                    handleAddNestedCategory={this.props.handleAddNestedCategory}
                    key={category.name}
                    handleEditCategory={this.props.handleEditCategory}
                    handleDeleteCategory={this.props.handleDeleteCategory}
                    mode={this.props.mode}
                    updateProgressBar={this.props.updateProgressBar}
                    categoryID={this.props.categoryID}
                    handleSpecifyCategoryToMove={this.handleSpecifyCategoryToMove}
                    categoryToMoveTaskInId={this.state.categoryToMoveTaskInId}
                    transferTaskIn={this.state.categoryToMoveTaskInId === category.id}
                />
            );
        });

        return (
            <div className="categories__container">
                <ul className="categories__list">
                    {categories}
                </ul>
            </div>
        );
    }
}

export default Categories;
