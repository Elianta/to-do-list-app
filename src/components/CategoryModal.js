import React from 'react';
import Modal from 'react-modal';

const CategoryModal = (props) => (
    <Modal
        isOpen={props.categoryModalDisplayed}
        onRequestClose={props.handleModalClose}
        contentLabel="Selected Category"
        closeTimeoutMS={200}
        className="modal"
    >
        <h3 className="modal__title">{props.title}</h3>
        <form className="modal__form" onSubmit={props.handleSaveCategory}>
            <input
                type="text"
                name="category"
                autoFocus={true}
                className="modal__input"
                defaultValue={props.modalMode === 'edit' ? props.categoryName : ''}
            />
            <button className="button  modal__btn">Save</button>
        </form>
    </Modal>
);

export default CategoryModal;
