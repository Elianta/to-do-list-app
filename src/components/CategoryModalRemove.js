import React from 'react';
import Modal from 'react-modal';

const CategoryModalRemove = (props) => (
    <Modal
        isOpen={props.categoryModalRemoveDisplayed}
        onRequestClose={props.handleModalClose}
        contentLabel="Selected Category"
        closeTimeoutMS={200}
        className="modal"
    >
        <h3 className="modal__title">{`Do you want to remove '${props.categoryName}'?`}</h3>
        <p className="modal__content">(All nested categories will be also removed)</p>
        <button className="button  modal__btn" onClick={props.handleRemoveConfirm}>OK</button>
        <button className="button  modal__btn" onClick={props.handleModalClose}>Cancel</button>
    </Modal>
);

export default CategoryModalRemove;
