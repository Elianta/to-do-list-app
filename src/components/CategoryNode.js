export default class CategoryNode {
  constructor(name, id) {
    this.name = name;
    this.id = id;
    this.children = [];
    this.tasks = [];
  }
}