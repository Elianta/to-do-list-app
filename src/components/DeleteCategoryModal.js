import React from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import Modal from 'react-modal';
import {removeCategory} from '../actions/categories';
import {closeModals} from '../actions/modals';

export class DeleteCategoryModal extends React.Component {
    constructor(props) {
        super(props);
        this.handleDeleteCategory = this.handleDeleteCategory.bind(this);
    }

    handleDeleteCategory(e) {
        e.preventDefault();
        this.props.removeCategory(this.props.modals.editableCategory.id);
        this.props.closeModals();
    }

    render() {
        return (
            <Modal
                isOpen={this.props.modals.isDeleteCategoryModalOpen}
                onRequestClose={this.props.handleModalClose}
                contentLabel="Delete Category"
                closeTimeoutMS={200}
                className="modal"
            >
                <h3 className="modal__title">{`Do you want to delete '${this.props.modals.editableCategory.name}'?`}</h3>
                <p className="modal__content">(All nested categories will be also removed)</p>
                <button data-testid="confirm-btn" className="button  modal__btn" onClick={this.handleDeleteCategory}>OK</button>
                <button data-testid="cancel-btn" className="button  modal__btn" onClick={this.props.handleModalClose}>Cancel</button>
            </Modal>
        );
    }
}

const mapStateToProps = (state) => ({
    modals: state.modals
});

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
        removeCategory,
        closeModals
    }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(DeleteCategoryModal);
