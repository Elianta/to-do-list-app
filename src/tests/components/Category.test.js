import React from 'react';
import {shallow} from 'enzyme';
import {Category} from '../../components/Category';

describe('taskTransfer mode', () => {
    const node = {
        id: 303,
        name: 'Category Name #1-22-33'
    };
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <Category
                mode={'taskTransfer'}
                node={node}
                children={[]}
                active={'303'}
                transferTaskIn={false}
            />
        );
    });
    test('should render Category in the taskTransfer mode', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('transfer button should be disabled for selected directory', () => {
        const transferBtn = wrapper.find('[data-testid="transfer-btn"]');
        expect(transferBtn.prop('disabled')).toBeTruthy();
    });
});

describe('homePage mode', () => {
    const node = {
        id: 12,
        name: 'Category Name #12'
    };
    const openModalForCategoryEditing = jest.fn();
    const openModalForCategoryDeleting = jest.fn();
    const openModalForCategoryAddition = jest.fn();
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <Category
                mode={'homePage'}
                node={node}
                children={[]}
                active={'12'}
                transferTaskIn={false}
                openModalForCategoryEditing={openModalForCategoryEditing}
                openModalForCategoryDeleting={openModalForCategoryDeleting}
                openModalForCategoryAddition={openModalForCategoryAddition}
            />
        );
    });

    test('should render Category in the homePage mode', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should handle click on edit button correctly', () => {
        const editBtn = wrapper.find('[data-testid="edit-btn"]');
        editBtn.simulate('click');
        expect(wrapper.instance().props.openModalForCategoryEditing).toHaveBeenCalledWith({id: node.id, name: node.name});
    });

    test('should handle click on delete button correctly', () => {
        const deleteBtn = wrapper.find('[data-testid="delete-btn"]');
        deleteBtn.simulate('click');
        expect(wrapper.instance().props.openModalForCategoryDeleting).toHaveBeenCalledWith({id: node.id, name: node.name});
    });

    test('should handle click on add button correctly', () => {
        const addBtn = wrapper.find('[data-testid="add-btn"]');
        addBtn.simulate('click');
        expect(wrapper.instance().props.openModalForCategoryAddition).toHaveBeenCalledWith({id: node.id, name: node.name});
    });
});


