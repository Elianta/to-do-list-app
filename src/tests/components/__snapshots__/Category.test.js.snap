// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`homePage mode should render Category in the homePage mode 1`] = `
<li
  className="categories__item"
  key="12"
>
  <NavLink
    activeClassName="categories__link--active"
    ariaCurrent="true"
    className="categories__link"
    exact={true}
    isActive={[Function]}
    to="/category/12"
  >
    Category Name #12
  </NavLink>
  <div
    className="categories__manage-list"
  >
    <button
      className="button  button--manage  categories__manage-item"
      data-testid="edit-btn"
      onClick={[Function]}
    >
      <svg
        height="20"
        viewBox="0 0 528.899 528.899"
        width="20"
      >
        <path
          d="M328.883,89.125l107.59,107.589l-272.34,272.34L56.604,361.465L328.883,89.125z M518.113,63.177l-47.981-47.981 c-18.543-18.543-48.653-18.543-67.259,0l-45.961,45.961l107.59,107.59l53.611-53.611 C532.495,100.753,532.495,77.559,518.113,63.177z M0.3,512.69c-1.958,8.812,5.998,16.708,14.811,14.565l119.891-29.069 L27.473,390.597L0.3,512.69z"
        />
      </svg>
    </button>
    <button
      className="button  button--manage  categories__manage-item"
      data-testid="delete-btn"
      onClick={[Function]}
    >
      <svg
        height="20"
        viewBox="0 0 408.483 408.483"
        width="20"
      >
        <path
          d="M87.748,388.784c0.461,11.01,9.521,19.699,20.539,19.699h191.911c11.018,0,20.078-8.689,20.539-19.699l13.705-289.316 H74.043L87.748,388.784z M247.655,171.329c0-4.61,3.738-8.349,8.35-8.349h13.355c4.609,0,8.35,3.738,8.35,8.349v165.293 c0,4.611-3.738,8.349-8.35,8.349h-13.355c-4.61,0-8.35-3.736-8.35-8.349V171.329z M189.216,171.329 c0-4.61,3.738-8.349,8.349-8.349h13.355c4.609,0,8.349,3.738,8.349,8.349v165.293c0,4.611-3.737,8.349-8.349,8.349h-13.355 c-4.61,0-8.349-3.736-8.349-8.349V171.329L189.216,171.329z M130.775,171.329c0-4.61,3.738-8.349,8.349-8.349h13.356 c4.61,0,8.349,3.738,8.349,8.349v165.293c0,4.611-3.738,8.349-8.349,8.349h-13.356c-4.61,0-8.349-3.736-8.349-8.349V171.329z"
        />
        <path
          d="M343.567,21.043h-88.535V4.305c0-2.377-1.927-4.305-4.305-4.305h-92.971c-2.377,0-4.304,1.928-4.304,4.305v16.737H64.916 c-7.125,0-12.9,5.776-12.9,12.901V74.47h304.451V33.944C356.467,26.819,350.692,21.043,343.567,21.043z"
        />
      </svg>
    </button>
    <button
      className="button  button--manage"
      data-testid="add-btn"
      onClick={[Function]}
    >
      <svg
        height="22"
        viewBox="0 0 33.059 33.059"
        width="22"
      >
        <path
          d="M15.529,31.059C6.966,31.059,0,24.092,0,15.529C0,6.966,6.966,0,15.529,0 c8.563,0,15.529,6.966,15.529,15.529C31.059,24.092,24.092,31.059,15.529,31.059z M15.529,1.774 c-7.585,0-13.755,6.171-13.755,13.755s6.17,13.754,13.755,13.754c7.584,0,13.754-6.17,13.754-13.754S23.113,1.774,15.529,1.774z"
        />
        <path
          d="M21.652,16.416H9.406c-0.49,0-0.888-0.396-0.888-0.887c0-0.49,0.397-0.888,0.888-0.888h12.246 c0.49,0,0.887,0.398,0.887,0.888C22.539,16.02,22.143,16.416,21.652,16.416z"
        />
        <path
          d="M15.529,22.539c-0.49,0-0.888-0.397-0.888-0.887V9.406c0-0.49,0.398-0.888,0.888-0.888 c0.49,0,0.887,0.398,0.887,0.888v12.246C16.416,22.143,16.02,22.539,15.529,22.539z"
        />
      </svg>
    </button>
  </div>
</li>
`;

exports[`taskTransfer mode should render Category in the taskTransfer mode 1`] = `
<li
  className="categories__item"
  key="303"
>
  <span
    className="categories__text"
  >
    Category Name #1-22-33
  </span>
  <div
    className="categories__manage-list"
  >
    <button
      className="button  button--manage  button--transfer"
      data-testid="transfer-btn"
      disabled={true}
      onClick={[Function]}
    >
      <svg
        height="20"
        viewBox="0 0 533.333 533.333"
        width="20"
      >
        <path
          d="M241.914,403.974v129.359l-200-200l200-200v132.2C474.597,270.992,464.616,107.309,405.384,0 C551.587,158.028,520.541,411.234,241.914,403.974z"
        />
      </svg>
    </button>
  </div>
</li>
`;
